UI: Add setting to copy screenshots to clipboard

jpark37 opened the pull request:
### Description
Add ability to save screenshots to the OS clipboard.

HDR images will be converted to SDR.

Skip Linux because the image is sometimes corrupt when tested on Ubuntu.

### Motivation and Context
Useful for working with programs that accept images from the clipboard.

### How Has This Been Tested?
- [x] Windows: Clipboard is not used when setting is disabled
- [x] Windows, HDR: Verified clipboard matches file and is correct
- [x] Windows, SDR: Verified clipboard is SDR tonemapped compared to JXR file and both are correct
- [x] Mac: Clipboard is not used when setting is disabled
- [x] Mac, SDR: Verified clipboard matches file and is correct
- [x] Mac, HDR: Verified clipboard is correct, and SDR file hasn't regressed
- [x] Linux: Setting is not visible
- [x] Linux: Screenshot to file still works

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Warchamp7 commented:
Very much on board with screenshots going to the clipboard, but I'd like to request some UX adjustments

1. Do not add a new context menu entry
2. Add an option in Advanced -> Recording section for 'Copy screenshots to clipboard'
2b. Probably rename the Recording section of Advanced to 'Output', as the settings here apply to Replay Buffer and Screenshots as well these days. Might be better as it's own PR
2c. Default 'Copy screenshots to clipboard' to on/checked

By default screenshots should be saved to file, and placed on the clipboard when supported, with an option to disable copying to clipboard.

jpark37 commented:
I'm fine with that except defaulting the setting to on. A user shouldn't expect to lose the existing contents of their clipboard just because they took a screenshot. If they opt in, then it's on them.

Personally, I don't want image files being written, but I'm happy to have this feature in any form.

Warchamp7 commented:
> I'm fine with that except defaulting the setting to on. A user shouldn't expect to lose the existing contents of their clipboard just because they took a screenshot. If they opt in, then it's on them.
> 
> Personally, I don't want image files being written, but I'm happy to have this feature in any form.

On Windows at least, the built in Prnt Scr system hotkey ONLY goes to clipboard, and their Snipping Tool defaults to copying to clipboard on both Windows 10 and 11. I would never propose copying screenshots to the clipboard without an option to disable it, but I think there is reasonable precedent for it as the default (again, on Windows at least)

Speaking personally I was actually surprised (and frustrated 🥲) that OBS screenshots _didn't_ go to the clipboard when the feature was originally added, heh

jpark37 committed:
UI: Add setting to copy screenshots to clipboard

HDR images will be converted to SDR.

Skip Linux because the image is sometimes corrupt when tested on Ubuntu.

jpark37 commented:
I made the changes although I kept the default off.

Print Screen and Snipping Tool are copying to clipboard as the primary purpose of their actions, not as a side effect that the user doesn't expect. Messing with the clipboard of someone who is just trying to save a file without their consent is a bit aggressive.

jpark37 commented:
On, snipping tool saves files to `Pictures\Screenshots`. Sorry, never used it before.

I'm still on the fence... for my use case I'd want it on all the time, but I don't want to take the heat for blowing up users' clipboards, especially since we're changing the behavior of an existing action.

Warchamp7 commented:
> On, snipping tool saves files to `Pictures\Screenshots`. Sorry, never used it before.
> 
> I'm still on the fence... for my use case I'd want it on all the time, but I don't want to take the heat for blowing up users' clipboards, especially since we'd be changing the behavior of an existing action.

If anyone comes to you upset about it you are beyond welcome to direct them to me ;P

jpark37 commented:
Ok, it's on by default.

PatTheMav commented:
> I'm fine with that except defaulting the setting to on. A user shouldn't expect to lose the existing contents of their clipboard just because they took a screenshot. If they opt in, then it's on them.
> 
> Personally, I don't want image files being written, but I'm happy to have this feature in any form.

FWIW I argued for the same, so thank you for pointing that out as well. 😄

ryantheleach commented:
Personally I wouldn't want old temporary debugging screenshots hogging up diskspace if I intend on using the copy to clipboard functionality. If copy to clipboard is an option, then saving to disk must also be IMO.
